src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+            _localEscapeScopes[local] = (refEscapeScope, valEscapeScope);
+        }
+
+        private void AddPlaceholderScopes(ArrayBuilder<(BoundValuePlaceholderBase, uint)> placeholders)
+        {
+            _placeholderScopes ??= new Dictionary<BoundValuePlaceholderBase, uint>();
+            foreach (var (placeholder, valEscapeScope) in placeholders)
+            {
+                _placeholderScopes.Add(placeholder, valEscapeScope);
+            }
+        }
+
+#pragma warning disable IDE0060
+        private void RemovePlaceholderScopes(ArrayBuilder<BoundValuePlaceholderBase> placeholders)
+        {
+            Debug.Assert(_placeholderScopes is { });
+            // https://github.com/dotnet/roslyn/issues/65961: Currently, analysis may require subsequent calls
+            // to GetRefEscape(), etc. for the same expression so we cannot remove placeholders eagerly.
+            //foreach (var placeholder in placeholders)
+            //{
+            //    _placeholderScopes.Remove(placeholder);
+            //}
+        }
+#pragma warning restore IDE0060
+
+        private uint GetPlaceholderScope(BoundValuePlaceholderBase placeholder)
+        {
+            Debug.Assert(_placeholderScopes is { });
+            return _placeholderScopes[placeholder];
+        }
+
+        public override BoundNode? VisitBlock(BoundBlock node)
+        {
+            var inUnsafeRegion = node.InUnsafeRegion;
+            using var region = new UnsafeRegion(this, inUnsafeRegion.HasValue() ? inUnsafeRegion.Value() : _inUnsafeRegion);
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitBlock(node);
+        }
+
+        public override BoundNode? Visit(BoundNode? node)
+        {
+#if DEBUG
+            if (node is BoundValuePlaceholderBase placeholder
+                && node is not (BoundObjectOrCollectionValuePlaceholder or BoundInterpolatedStringHandlerPlaceholder))
+            {
+                Debug.Assert(_placeholderScopes?.ContainsKey(placeholder) == true);
+            }
+#endif
+            return base.Visit(node);
+        }
+
+        private void VisitFieldOrPropertyInitializer(BoundInitializer initializer)
+        {
+            var fieldEqualsValue = (BoundFieldEqualsValue)initializer;
+
+            using var _ = new LocalScope(this, fieldEqualsValue.Locals);
+
+            base.Visit(fieldEqualsValue.Value);
+
+            var field = fieldEqualsValue.Field;
+            bool isByRef = field.RefKind != RefKind.None;
+            if (isByRef || field.Type.IsRefLikeType)
+            {
+                ValidateEscape(fieldEqualsValue.Value, Binder.CallingMethodScope, isByRef: isByRef, _diagnostics);
+            }
+        }
+
+        public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
+        {
+            var localFunction = node.Symbol;
+            // https://github.com/dotnet/roslyn/issues/65353: We should not reuse _localEscapeScopes
```

Comment:
I assume this is also not observable. Is that right?
nit: Is there some complication that motivates not addressing in current PR?

Reply:
This issue is observable - see the examples in [https://github.com/dotnet/roslyn/issues/65353](https://github.com/dotnet/roslyn/issues/65353).

We could try fixing in this PR, although since it is an existing issue, it seemed perhaps simpler to address separately.